Add download files

tommyv1987 opened the pull request:
## Ticket

## Description

- add retry logic for downloads/pings with early exit after 3 consecutive failures
- include new download urls
- inroduce better timeout handling to reduce wasted time on broken connections
- add tests
- change default ipv4 address for pinging (nymtech.net has a redirect to nym.com)
- change ipv6.google.com to cloudfare 
- added some basic tests around the netstack args (more to come)

## Checklist:

- [ ] Changelog

## Screenshots (optional, if UI related)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594)
<!-- Reviewable:end -->

tommyv1987 committed:
Improve download reliability with retry logic

- add retry logic for downloads/pings with early exit after 3 consecutive failures
- include new download urls
- inroduce better timeout handling to reduce wasted time on broken connections.

tommyv1987 committed:
Actually add the logic for early exit

tommyv1987 committed:
Run mod tidy too - to validate

tommyv1987 committed:
Add tests for gateway probe retry logic

- add tests for retry and helper functions
- note: TestPingFunction requires proper WireGuard network setup to pass

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594)* status: 0 of 2 files reviewed, 2 unresolved discussions

___
*[`nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go` line 227 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594#-OaVkK-x1XMbanUZ9NZa:-OaVkK-x1XMbanUZ9NZb:b-rd2he8) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/79303e83dcbe8f0c146c1587ed142c613a17ded6/nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go#L227)):*
> ```Go
> 					if consecutiveFailures >= maxConsecutiveFailures {
> 						log.Printf("Too many consecutive failures (%d), stopping ping attempts for %s", consecutiveFailures, ip)
> 						return
> ```

Does this return break the loop given that it's within the closure? cc @jmwample 
___
*[`nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go` line 229 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594#-OaVkahE3k2SWoml0WiI:-OaVkahE3k2SWoml0WiJ:b-31t020) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/79303e83dcbe8f0c146c1587ed142c613a17ded6/nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go#L229)):*
> ```Go
> 						return
> 					}
> 					return
> ```

You seem to return regardless :shrug: 


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @jmwample and @tommyv1987)

___
*[`nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go` line 384 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594#-OaVl04t4cECOohmP1wc:-OaVl04u-e8DHNaO3Z4r:b-gtisxv) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/79303e83dcbe8f0c146c1587ed142c613a17ded6/nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go#L384)):*
> ```Go
> }
> 
> func abs(x int) int {
> ```

Maybe you can use `math.Abs(value)`, example:

```go
package main 

import ("fmt" "math") 
func main() { 
  x := -25.0 
  y := math.Abs(x) 
  fmt.Print(y) 
}
```


<!-- Sent from Reviewable.io -->

pronebird reviewed (commented):
> @pronebird reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594)* status: all files reviewed, 3 unresolved discussions (waiting on @jmwample and @tommyv1987)

___
*[`nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go` line 229 at r1](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594#-OaVkahE3k2SWoml0WiI:-OaVmC3w0Pwju6tDuidU:bffj2po) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/79303e83dcbe8f0c146c1587ed142c613a17ded6/nym-vpn-core/crates/nym-gateway-probe/netstack_ping/lib.go#L229)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

You seem to return regardless :shrug:
</blockquote></details>

Maybe extract the body and remove the closure around `the func() { /* body */ }()` and add  `time.Sleep(5 * time.Second)` instead of this return to sleep a bit after each iteration


<!-- Sent from Reviewable.io -->

tommyv1987 committed:
Address PR feedback and improve ping logic

- added additional test to check breakout loop
- replace custom abs() with math.Abs()

tommyv1987 commented:
Successful output of the node after test run:
```
{
  "node": "2KuFi7CjKVh1mQrNKhdwLjhXBhVLRktvWMg5k4WXKdrX",
  "used_entry": "2KuFi7CjKVh1mQrNKhdwLjhXBhVLRktvWMg5k4WXKdrX",
  "outcome": {
    "as_entry": {
      "can_connect": true,
      "can_route": true
    },
    "as_exit": {
      "can_connect": true,
      "can_route_ip_v4": true,
      "can_route_ip_external_v4": true,
      "can_route_ip_v6": true,
      "can_route_ip_external_v6": true
    },
    "wg": {
      "can_register": true,
      "can_handshake_v4": true,
      "can_resolve_dns_v4": true,
      "ping_hosts_performance_v4": 1.0,
      "ping_ips_performance_v4": 1.0,
      "can_handshake_v6": true,
      "can_resolve_dns_v6": true,
      "ping_hosts_performance_v6": 1.0,
      "ping_ips_performance_v6": 0.93333334,
      "download_duration_sec_v4": 3,
      "downloaded_file_v4": "https://proof.ovh.net/files/1Mb.dat",
      "download_error_v4": "",
      "download_duration_sec_v6": 3,
      "downloaded_file_v6": "https://nym-bandwidth-monitoring.ops-d86.workers.dev/10mb.dat",
      "download_error_v6": ""
    }
  }
}
```

tommyv1987 committed:
Change a few default args and validate netstack args

- nymtech.net has been replaced as it's not really relevant anymore as nym.com exists
- switch out to use cloudfare too

pronebird reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @pronebird reviewed 2 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/3594#-:-OaZJ3261wNzR-nFuSTi:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @tommyv1987)



<!-- Sent from Reviewable.io -->
